providers/azureblob/src/test/java/org/jclouds/azure/storage/util/StorageUrlDelegateTest.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jclouds.azure.storage.util;
+
+import static org.testng.Assert.assertEquals;
+
+import java.util.Properties;
+
+import org.jclouds.ContextBuilder;
+import org.jclouds.azure.storage.reference.AzureConstants;
+import org.jclouds.logging.config.NullLoggingModule;
+import org.jclouds.rest.internal.BaseRestApiTest;
+import org.testng.annotations.Test;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Module;
+
+@Test(groups = "unit")
+public class StorageUrlDelegateTest {
+
+   private static final String ACCOUNT = "foo";
+   private static final ImmutableSet<Module> MODULES = ImmutableSet.<Module> of(new BaseRestApiTest.MockModule(),
+         new NullLoggingModule());
+
+   @Test
+   void testDefaultEndpoint() {
+
+      StorageUrlDelegate target = ContextBuilder
+            .newBuilder("azureblob")
+            .credentials(ACCOUNT, "?token")
+            .modules(MODULES)
+            .buildInjector().getInstance(StorageUrlDelegate.class);
+
+      assertEquals(target.configureStorageUrl(), "https://foo.blob.core.windows.net/");
+   }
+
+   @Test
+   void testCustomEndpointWithoutStorageAccountPath() {
+
+      StorageUrlDelegate target = ContextBuilder
+            .newBuilder("azureblob").endpoint("http://localhost:10000")
+            .credentials(ACCOUNT, "?token")
+            .modules(MODULES)
+            .buildInjector()
+            .getInstance(StorageUrlDelegate.class);
+
+      assertEquals(target.configureStorageUrl(), "http://localhost:10000/");
```

Comment:
Why are we only setting the account name if `PROPERTY_AZURE_VIRTUAL_HOST_STORAGE_ACCOUNT == false` is false? I mean, what is the point in not adding the account name anywhere in the URI? Either if we have a custom endpoint or not, I guess we want the account name to be present _somewhere_?

Reply:
For azure, it's already present in the vhost/subdomain, having it in the path too would be incorrect.  Therefore it must be configurable as the behaviour of the azurite endpoint looks for it in the path.

Reply:
My point is, does it make sense to _ignore_ the account if a custom endpoint is configured? Should we just do the same as we do when `PROPERTY_AZURE_VIRTUAL_HOST_STORAGE_ACCOUNT == false` when a custom endpoint is set (regardless of the property value)?

Which leads me to think: doe sit make sense to set the property to `vhost` if we're setting a custom endpoint? because the current code does not handle that case. And if it does not make sense, do we still need the property? Can we just default to vhost, but if there is a custom endpoint, just append the account to the path?

Reply:
Possibly. I was following the pattern used by the AWS S3 connector which has a similar property.

But I'm not sure there's any harm in having it configurable instead of automatic?

Reply:
It is about correctness. If there is a custom endpoint and the property has the default value, the account part never manifests in the generated storage URL. Is that correct?

Reply:
There are use cases where it is be desirable to have a custom endpoint without activating the storage-account-in-path format.

For example you set up a custom DNS name for your blob storage endpoint: 
https://docs.microsoft.com/en-us/azure/storage/blobs/storage-custom-domain-name?tabs=azure-portal

This would not be running the local service but you would be using a custom endpoint.

This is not a use case that I am interested in, but just throwing it out there.

Reply:
That makes sense, thanks for the link!

One last thing, then, if there is no custom endpoint configured and the property is set to `false`, the account will be included in the hostname, but appended to the path as well, which I guess is not correct?

Reply:
No worries, I appreciate the comments.

Sure, it defaults to "true". If the behaviour is explicitly configured to "false" in the scenario you've outlined, then you are right this would not be correct.

I don't think it's possible to prevent misconfigurations like this? I would not expect most users to override the default behaviour unless they had a reason to?

Did you have any suggestions?

Reply:
The problem here is that given random endpoints, we can't preconfigure a correct way of setting the account for every possible case. What bout this:

Let's change the `StorageUrlDelegate` to an interface like this:
```java
@ImplementedBy(StorageAccountInVhost.class)
public interface StorageURLSupplier extends Supplier<URI> {
}
```
And have the default implementation be `StorageAccountInVhost`, which just keeps the current behavior, with the hardcoded endpoint and the identity in the host.

Then we could have a:
```java
public class AppendAccountToEndpoint implements StorageURLSupplier
```
That just appends the identity to whatever endpoint is configured.

To make it easier to configure, we could have a Guice module as well that does this (but is not included by default in the azure blob modules list:
```java
public class AppendAccountToEndpointModule extends AbstractModule {
   @Override
   protected void configure() {
      bind(StorageURLSupplier.class).to(AppendAccountToEndpoint.class);
   }
}
```
With these three pieces there, you can plug in your storage URL strategy when creating the context by adding the `AppendAccountToEndpointModule` to the list of modules.

This way we don't need the property, achieve the same behavior, but we also allow users to customize this further, as they can provide their own module to configure how the identity should be added to the endpoint.
This approach may seem overkill, but it ensures correctness in configuration.

WDYT?

Reply:
Sounds good, I've refactored the solution to follow your approach.  Thanks.